native/dgebal.go

```diff
@@ -0,0 +1,227 @@
+// nonzero off-diagonal element in each row and column. The indices ilo and ihi
+// mark the starting and ending columns of the submatrix B. The eigenvalues of A
+// isolated in the first 0 to ilo-1 and last ihi+1 to n-1 elements on the
+// diagonal can be read off without any roundoff error.
+//
+// Scaling consists of applying a diagonal similarity transformation D such that
+// D^{-1}*B*D has the 1-norm of each row and its corresponding column nearly
+// equal. The output matrix is
+//  [ T1     X*D          Y    ]
+//  [  0  inv(D)*B*D  inv(D)*Z ].
+//  [  0      0           T2   ]
+// Scaling may reduce the 1-norm of the matrix, and improve the accuracy of
+// the computed eigenvalues and/or eigenvectors.
+//
+// job specifies the operations that will be performed on A.
+// If job is lapack.None, Dgebal sets scale[i] = 1 for all i and returns ilo=0, ihi=n-1.
+// If job is lapack.Permute, only permuting will be done.
+// If job is lapack.Scale, only scaling will be done.
+// If job is lapack.PermuteScale, both permuting and scaling will be done.
+//
+// On return, if job is lapack.Permute or lapack.PermuteScale, it will hold that
+//  A[i,j] == 0,   for i > j and j∈{0, ..., ilo-1, ihi+1, ..., n-1}.
+// If job is lapack.None or lapack.Scale, or if n == 0, it will hold that
+//  ilo == 0 and ihi == n-1.
+//
+// On return, scale will contain information about the permutations and scaling
+// factors applied to A. If π(j) denotes the index of the column interchanged
+// with column j, and D[j,j] denotes the scaling factor applied to column j,
+// then
+//  scale[j] == π(j),     for j∈{0, ..., ilo-1, ihi+1, ..., n-1},
+//           == D[j,j],   for j∈{ilo, ..., ihi}.
+//
+// Dgebal is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dgebal(job lapack.Job, n int, a []float64, lda int, scale []float64) (ilo, ihi int) {
+	switch job {
+	default:
+		panic(badJob)
+	case lapack.None, lapack.Permute, lapack.Scale, lapack.PermuteScale:
+	}
+	checkMatrix(n, n, a, lda)
+	if len(scale) != n {
+		panic("lapack: bad length of scale")
+	}
+
+	ilo = 0
+	ihi = n - 1
+
+	if n == 0 || job == lapack.None {
+		for i := range scale {
+			scale[i] = 1
+		}
+		return ilo, ihi
+	}
+
+	bi := blas64.Implementation()
+	swapped := true
+
+	if job == lapack.Scale {
+		goto Scaling
+	}
+
+	// Permutation to isolate eigenvalues if possible.
+	//
+	// Search for rows isolating an eigenvalue and push them down.
+	for swapped {
+		swapped = false
+	RowLoop:
+		for i := ihi; i >= 0; i-- {
+			for j := 0; j <= ihi; j++ {
+				if i == j {
+					continue
+				}
+				if a[i*lda+j] != 0 {
+					continue RowLoop
+				}
+			}
+			// Row i has only zero off-diagonal elements in the
+			// block A[ilo:ihi+1,ilo:ihi+1].
+			scale[ihi] = float64(i)
+			if i != ihi {
+				bi.Dswap(ihi+1, a[i:], lda, a[ihi:], lda)
+				bi.Dswap(n, a[i*lda:], 1, a[ihi*lda:], 1)
+			}
+			ihi--
+			if ihi == 0 {
+				scale[0] = 1
```

Comment:
Fortran here is very subtle. They don't set the scale explicitly, they do one extra iteration with ihi=1 and rely on the fact that that value of ihi is also the right value for scale[ihi].